### ISSUE TITLE
Reintroducing unnecessary modifiers not detected unless other issues present in same file

### ISSUE THREAD

**Issue by shrugalic** [solven-eu/cleanthat#802]
# Setup
We just added `cleanthat` to our Gradle build of a Java 17 project, via [`spotless`](https://github.com/diffplug/spotless/tree/main/plugin-gradle#cleanthat), with default settings:
```kts
cleanthat().sourceCompatibility("17")
```
By doing a `./gradlew spotlessApply` It nicely cleaned up various things, including unnecessary modifiers in interfaces:
```diff
interface Test {
+    String "TEST";
-    public static final String "TEST";
}
```
As well as issues like this one:
```diff
List<String> list = List.of("test");
-if (list.size() == 0) {
+if (list.isEmpty()) {
```

So far, so good.

# Issue
To check if `./gradlew spotlessCheck` would detect any `cleanthat` issues, we reverted the first of the above changes – that is, we reintroduced the unnecessary modifiers of the constants within the interface.

`spotlessCheck` did not detect any issues, nor did `spotlessApply` fix it.

However, if we also reintroduced the second issue (`list.size() == 0`) **in the same file**, then `spotlessCheck` did detect the issue, and `spotlessApply` did fix it.

If the second issue was reintroduced into **an other** file, it would detect only that issue, and continue to ignore the unnecessary modifiers in the other file.

# Expectation
It would be great if unnecessary modifiers would be detected even if no other issues are present in a file.

**Comment by blacelle** [comment]
I feel like this is a spotless effect (maybe not issue, but limitation), and not specific to spotless. Typically related to `ratchetFrom` option, and detection of which files are to be considered.

Is this reproducible in an open-source project?

**Comment by shrugalic** [comment]
> I feel like this is a spotless effect (maybe not issue, but limitation), and not specific to spotless.

Did you maybe mean _not specific to **cleanthat**_?


> Is this reproducible in an open-source project?

Hmm, I don't have one at hand. Did you mean any random Java open source project using Gradle and thus calling `cleanthat` via `spotless`, rather than by other means?

**Comment by blacelle** [comment]
> > I feel like this is a spotless effect (maybe not issue, but limitation), and not specific to spotless.

> Did you maybe mean not specific to cleanthat?

yes

---

>  Did you mean any random Java open source project using Gradle and thus calling cleanthat via spotless, rather than by other means?

Any clearer reproduction case would be helpful. You provided a scenario with OK steps and KO steps.

> spotlessCheck did not detect any issues, nor did spotlessApply fix it.

I understand `spotlessApply` did not cleaned as expected a `.java` file with only unnecessary modifiers.

> If the second issue was instead reintroduced into another (second) file, it would detect only that issue, and continue to ignore the unnecessary modifiers in the first file, as long as that issue was the only on there.

Here, I underatand `spotlessApply` did not cleaned as expected a `.java` file with unnecessary modifiers given other issues.

Could you provide a `.java` file which is not cleaned as expected?

(Your report suggest the rules kind of conflicts, which would be an unusual bug).

**Comment by shrugalic** [comment]
> Could you provide a `.java` file which is not cleaned as expected?

Sure.

# How to reproduce
The following interface will not be touched by `spotlessCheck` or `spotlessApply`, even though it has unnecessary `public static final` modifiers:
```java
public interface TopLevelInterface {
    public static final String TEST = "value";
}
```
Adding another issue to the same file will result in both issues being fixed. Here's file before `spotlessApply`:
```java
import java.util.List;

public interface TopLevelInterface {
    public static final String TEST = "value";

    default void method() {
        if (List.of().size() == 0) {
            // nop
        }
    }
}
```
Output of `./gradlew spotlessCheck` for this file:
```sh
* What went wrong:
Execution failed for task 'spotlessJavaCheck'.
> The following files had format violations:
      src/TopLevelInterface.java
          @@ -3,10 +3,10 @@
           import·java.util.List;

           public·interface·TopLevelInterface·{
          -\tpublic·static·final·String·TEST·=·"value";
          +\tString·TEST·=·"value";

           \tdefault·void·method()·{
          -\t\tif·(List.of().size()·==·0)·{
          +\t\tif·(List.of().isEmpty())·{
           \t\t\t//·nop
           \t\t}
           \t}
```
And here's the result of `./gradlew spotlessApply`:
```java
package ch.ergon.taifun.dis;

import java.util.List;

public interface TopLevelInterface {
    String TEST = "value";

    default void method() {
        if (List.of().isEmpty()) {
            // nop
        }
    }
}
```

If I were to add the unnecessary `public static final` modifiers back in, neither `spotlessCheck` nor `spotlessApply` will notice:
```java
package ch.ergon.taifun.dis;

import java.util.List;

public interface TopLevelInterface {
    public static final String TEST = "value";

    default void method() {
        if (List.of().isEmpty()) {
            // nop
        }
    }
}
```

**Comment by shrugalic** [comment]
>> If the second issue was instead reintroduced into another (second) file, it would detect only that issue, and continue to ignore the unnecessary modifiers in the first file, as long as that issue was the only on there.

> Here, I understand `spotlessApply` did not cleaned as expected a `.java` file with unnecessary modifiers given other issues.

Yes, but let me try to clarify. Assume we have two separate java files. One has unnecessary modifiers, and the other one has `List.of().size() == 0`. `spotlessCheck` will only see the latter file, and consequently `spotlessApply` will only fix the latter file. The issue in the first file is ignored.

A workaround would be to introduce an issue like `List.of().size() == 0` into the first tile, then both issues in this file would be fixed.

**Comment by blacelle** [comment]
I reproduce. No insight yet, but certainly not due to spotless.

**Comment by blacelle** [comment]
This ws actually a bug in the rule `ModifierOrder`, corrupting the AST around modifiers. This is fixed in 2.19.

**Comment by shrugalic** [comment]
I successfully verified that the issue was fixed, thanks a lot. 🙏
